Emit important replay frames on every judgement

bdach opened the pull request:
## [Emit important replay frames on every judgement](https://github.com/ppy/osu/commit/9f91c2e25ce76255dc39b90f5b1bfef3cfbfba99) 
- Closes https://github.com/ppy/osu/issues/4287
- Probably closes https://github.com/ppy/osu/issues/25405 (but not retroactively)
- Closes https://github.com/ppy/osu/issues/28677 (but not retroactively)

Up until now, whether or not a replay frame is emitted depended solely on the user's input, i.e. mouse movement or key presses/releases. This, intersected with the replay playback system which is given allowance to perform interpolation between replay frames, leads to potential situations wherein a replay can play inaccurately when a judgement takes place without user input meaningfully changing. One such case is slider ends with their 36ms of judgement leniency, see
https://github.com/ppy/osu/issues/25405#issuecomment-2879031106 for details on that; https://github.com/ppy/osu/issues/28677#issuecomment-2883534680 is another.

To that end, this commit aims to counteract that issue by *forcing* an important replay frame to be emitted on every new judgement recorded during gameplay. This will only benefit rulesets wherein judgements can occur that are not inherently tied to user input changing, which are going to be osu! as mentioned above, and ~~maybe possibly~~ catch. I don't foresee this doing anything relevant for taiko or mania.

## [Do not emit multiple consequent replay frames with the exact same data](https://github.com/ppy/osu/commit/cd4ab9307c5ed7d160919aa265392ecbad29a138)

With judgements occurring forcing emission of important frames, there's the question of emitting redundant frames. In a majority of cases still, a judgement *will* be tied to a specific user input, which means that there's a high likelihood of duplicate frames, as one will be emitted by the input change, and another by the judgement.

To a degree this is a pre-existing issue. The replay recording code responds to both mouse movement and key presses by recording frames, so depending on the intrinsic (basically undefined) ordering between handling mouse move and key presses, it was possible for multiple frames to be emitted with the same timestamp to the replay, each containing
partial input state changes (e.g. first frame with mouse position changed, and the second with a key pressed).

The extent to which this increases the frame count will obviously depend on the beatmap, but in my ballpark testing across all rulesets on a single "relatively standard" beatmap (think no aspire chicanery), around 4-10% of frames can end up being duplicates / not meaningful. This has implications both on replay size and on playback performance.

This commit therefore performs de-duplication of the frames based on timestamp - only the last frame with a given timestamp ends up in the final replay.

The CPU cost of this is negligible to the point of not being useful to profile - it's a single condition gated behind a single float comparison.

bdach committed:
Emit important replay frames on every judgement

- Closes https://github.com/ppy/osu/issues/4287
- Probably closes https://github.com/ppy/osu/issues/25405 (but not
  retroactively)

Up until now, whether or not a replay frame is emitted depended solely
on the user's input, i.e. mouse movement or key presses/releases. This,
intersected with the replay playback system which is given allowance to
perform interpolation between replay frames, leads to potential
situations wherein a replay can play inaccurately when a judgement takes
place without user input meaningfully changing. One such case is slider
ends with their 36ms of judgement leniency; see
https://github.com/ppy/osu/issues/25405#issuecomment-2879031106 for
details on that.

To that end, this commit aims to counteract that issue by *forcing* an
important replay frame to be emitted on every new judgement recorded
during gameplay. This will only benefit rulesets wherein judgements can
occur that are not inherently tied to user input changing, which are
going to be osu! as mentioned above, and maybe possibly catch. I don't
foresee this doing anything relevant for taiko or mania.

peppy commented on osu.Game/Rulesets/Replays/ReplayFrame.cs:
```diff
@@ -30,5 +30,10 @@ public ReplayFrame(double time)
+        /// <summary>
+        /// Whether this frame is equivalent to <paramref name="other"/> with respect to replay recording.
+        /// </summary>
+        public virtual bool IsEquivalentTo(ReplayFrame other) => Time == other.Time;
```
I'm honestly not sure there would be a case where time based comparison alone isn't enough, ie I don't know the overloads are required. But seems fine either way.

bdach replied:
I have seen such cases yes. Think move input and press input in a single frame. Both will emit replay frames but the press input may not yet be visible on the frame emitted via the move input. 

Whether deduplication of *those* frames should be done is a good question but one I didn't particularly want to broach in this PR because that is the pre-existing state but I can consider doing something about it in either this or a separate PR if you think it's worth it.

peppy replied:
This might be a good time to ensure sanity in such cases – one frame per timestamp – while we're here correcting mistakes. I believe the last frame pushed with same timestamp should take precedence (overwrite others).

I'm fine with that happening in this PR or in a follow-up, your choice.

bdach replied:
I'll see what I can do. I will note at this time that the most likely problem here is *spectator* - while for local recording I absolutely can replace the last frame with the same timestamp, it's going to be rather difficult to *un-send* the frame from spectator server even with the frame bundle buffering that we have.

So it would probably have to be a solution oriented towards *delaying* sending the frame until we're certain that time has indeed advanced.

peppy replied:
Right. I wouldn't care if the duplication still happens in spectator flow, but I can understand if that lenience doesn't reduce complexity.

bdach replied:
If I'm allowed to ignore spectator flow then something like https://github.com/ppy/osu/pull/33148/commits/cd4ab9307c5ed7d160919aa265392ecbad29a138 should do the job.

bdach replied:
There's one caveat here, namely https://github.com/ppy/osu/pull/33148/commits/953a93309c341edf733c669d3df5f4a6aa801c2a - if it ever turns out that it's possible that a key press and release can occur in a single frame of actual play, then this may have unintended consequences. I don't have evidence that this can occur from my testing, but that also doesn't mean that it *can't* occur.

> since future reports will be either for `.osr` replays or spectator realtime.

Not quite true either, it's going to be local `.osr` replays vs spectator realtime *or* server-side `.osr` replays.

peppy replied:
Hmm, that's something I completely forgot. We might want some similar logic server-side as well (or in `FrameDataBundle`? assuming frames don't get split across two frame bundles)

bdach replied:
> or in `FrameDataBundle`? assuming frames don't get split across two frame bundles

It would definitely be easiest / most convenient to do there, but there's a possibility of splitting across bundles as you say. Maybe I do that for now and we'll see what happens?

bdach committed:
Do not emit multiple consequent replay frames with the exact same data

With judgements occurring forcing emission of important frames, there's
the question of emitting redundant frames. In a majority of cases
still, a judgement *will* be tied to a specific user input, which means
that there's a high likelihood of duplicate frames, as one will be
emitted by the input change, and another by the judgement.

To a degree this is a pre-existing issue. The replay recording code
responds to both mouse movement and key presses by recording frames, so
depending on the intrinsic (basically undefined) ordering between
handling mouse move and key presses, it was possible for multiple frames
to be emitted with the same timestamp to the replay, each containing
partial input state changes (e.g. first frame with mouse position
changed, and the second with a key pressed).

The extent to which this increases the frame count will obviously depend
on the beatmap, but in my ballpark testing across all rulesets on a
single "relatively standard" beatmap (think no aspire chicanery), around
4-10% of frames can end up being duplicates / not meaningful. This has
implications both on replay size and on playback performance.

This commit therefore performs de-duplication of the frames based on
timestamp - only the last frame with a given timestamp ends up in the
final replay.

The CPU cost of this is negligible to the point of not being useful to
profile - it's a single condition gated behind a single float comparison

bdach committed:
Adjust failing replay recording tests

cd4ab9307c5ed7d160919aa265392ecbad29a138 made these tests fail because
they were basically pressing and releasing a key in a single frame,
which hopefully never actually happens in real scenarios. (It's a bit
tricky to determine if it can.)

bdach committed:
Also perform time-based frame de-duplication at frame bundle level

See https://github.com/ppy/osu/pull/33148#discussion_r2092467082

peppy committed:
Add matching comment in `SpectatorClient` for clarity

peppy reviewed (approved):
Let's see how this goes 😄
